### PR TITLE
RavenDB-18012 - [Sharding] Allow to get studio configuration (shard wide) and added EP abstraction

### PR DESCRIPTION
**Pull request by lastav5**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18012

### Additional description

Added processor to unify and prevent duplicate code of EPs that have same functionality in both sharded and regular.

### Type of change

- New feature

### How risky is the change?

- Low 

### Testing 

- It has been verified by manual testing

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Client/Documents/Operations/Configuration/GetStudioConfigurationOperation.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    public class GetStudioConfigurationOperation : IMaintenanceOperation<StudioConfiguration>
+    {
+        public RavenCommand<StudioConfiguration> GetCommand(DocumentConventions conventions, JsonOperationContext context)
+        {
+            return new GetStudioConfigurationCommand();
+        }
+
+        internal class GetStudioConfigurationCommand : RavenCommand<StudioConfiguration>
+        {
+            public override bool IsReadRequest => false;
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/Configuration/GetStudioConfigurationOperation.cs]: This _isn't_ a read request?

File: src/Raven.Server/Documents/ShardedHandlers/ShardedGetStudioConfigurationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿using System.Threading.Tasks;
+using Raven.Server.Documents.Handlers;
+using Raven.Server.Documents.Sharding;
+using Raven.Server.Routing;
+using Raven.Server.ServerWide.Context;
+
+namespace Raven.Server.Documents.ShardedHandlers
+{
+    internal class ShardedGetStudioConfigurationHandler : ShardedRequestHandler
```
Comment (ayende) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedGetStudioConfigurationHandler.cs]: Public?

Comment (lastav5) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedGetStudioConfigurationHandler.cs]: fixed

File: src/Raven.Client/Documents/Operations/Configuration/GetStudioConfigurationOperation.cs (hunk 1 above)
Comment (lastav5) [reply src/Raven.Client/Documents/Operations/Configuration/GetStudioConfigurationOperation.cs]: fixed

File: src/Raven.Server/Documents/Handlers/ConfigurationHandler.cs [hunk 3]
```diff
@@ -5,35 +5,57 @@
...
+using Raven.Server.Web.Studio.Sharding.Processors;
 using Sparrow.Json;
 
 namespace Raven.Server.Documents.Handlers
 {
     public class ConfigurationHandler : DatabaseRequestHandler
     {
-        [RavenAction("/databases/*/configuration/studio", "GET", AuthorizationStatus.ValidUser, EndpointType.Read)]
-        public async Task GetStudioConfiguration()
+        internal abstract class AbstractStudioConfigurationHandlerProcessor<TRequestHandler, TOperationContext> : AbstractHandlerProcessor<TRequestHandler, TOperationContext>
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/ConfigurationHandler.cs]: move this to a dedicated file inside the Processors namespace

File: src/Raven.Server/Documents/Handlers/ConfigurationHandler.cs [hunk 4]
```diff
@@ -5,35 +5,57 @@
...
+
+                    await using (var writer = new AsyncBlittableJsonTextWriter(context, RequestHandler.ResponseBodyStream()))
+                    {
+                        writer.WriteObject(clientConfigurationJson);
+                    }
                 }
             }
         }
 
+        internal class StudioConfigurationHandlerProcessor : AbstractStudioConfigurationHandlerProcessor<DatabaseRequestHandler, DocumentsOperationContext>
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/ConfigurationHandler.cs]: move this

File: src/Raven.Server/Documents/Handlers/ConfigurationHandler.cs [hunk 5]
```diff
@@ -5,35 +5,57 @@
...
+        {
+            public StudioConfigurationHandlerProcessor(DatabaseRequestHandler requestHandler, DocumentsContextPool contextPool) : base(requestHandler, contextPool) { }
+        }
+
+        [RavenAction("/databases/*/configuration/studio", "GET", AuthorizationStatus.ValidUser, EndpointType.Read)]
+        public async Task GetStudioConfiguration()
+        {
+            var configuration = Database.StudioConfiguration;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/ConfigurationHandler.cs]: Abstract processor should have abstract `GetStudioConfiguration` method and this should be moved there

File: src/Raven.Server/Documents/ShardedHandlers/ShardedGetStudioConfigurationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        public async Task GetStudioConfiguration()
+        {
+            var studioConfiguration = ShardedContext.DatabaseRecord.Studio;
+            using (var processor = new ShardedStudioConfigurationHandlerProcessor(this, ContextPool))
+            {
+                await processor.WriteStudioConfiguration(studioConfiguration);
+            }
+        }
+
+        internal class ShardedStudioConfigurationHandlerProcessor : ConfigurationHandler.AbstractStudioConfigurationHandlerProcessor<ShardedRequestHandler, TransactionOperationContext>
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedGetStudioConfigurationHandler.cs]: move this

File: src/Raven.Server/Documents/ShardedHandlers/ShardedGetStudioConfigurationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+using Raven.Server.ServerWide.Context;
+
+namespace Raven.Server.Documents.ShardedHandlers
+{
+    public class ShardedGetStudioConfigurationHandler : ShardedRequestHandler
+    {
+        [RavenShardedAction("/databases/*/configuration/studio", "GET")]
+        public async Task GetStudioConfiguration()
+        {
+            var studioConfiguration = ShardedContext.DatabaseRecord.Studio;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedGetStudioConfigurationHandler.cs]: should be in the impl of GetStudioConfiguration abstract method

File: src/Raven.Server/Documents/Handlers/ConfigurationHandler.cs [hunk 8]
```diff
@@ -5,35 +5,57 @@
...
+            { }
 
-            if (configuration == null)
-            {
-                HttpContext.Response.StatusCode = (int)HttpStatusCode.NotFound;
-                return;
-            }
-
-            using (ContextPool.AllocateOperationContext(out JsonOperationContext context))
+            public async Task WriteStudioConfiguration(StudioConfiguration studioConfiguration)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/ConfigurationHandler.cs]: not how it supposed to be.
add method
```
public async ValueTask ExecuteAsync();

protected StudioConfiguration GetStudioConfiguration();
```

**Pull request merged**